events/nats_connection.go

```diff
@@ -0,0 +1,150 @@
+package events
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"encoding/json"
+	"strings"
+
+	"github.com/nats-io/nats.go"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/trace"
+	"go.uber.org/zap"
+)
+
+const (
+	base10 = 10
+
+	natsTracerName = tracerName + ":nats"
+)
+
+var _ Connection = (*NatsConnection)(nil)
+
+// NatsConnection implements Connection.
+type NatsConnection struct {
+	logger    *zap.SugaredLogger
+	tracer    trace.Tracer
+	conn      *nats.Conn
+	jetstream nats.JetStreamContext
+	cfg       NatsConfig
+}
+
+// Shutdown gracefully drains the connection.
+func (c *NatsConnection) Shutdown(ctx context.Context) error {
+	ctx, cancelTimeout := context.WithTimeout(ctx, c.cfg.ShutdownTimeout)
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	defer cancelTimeout()
+
+	closedCB := c.conn.Opts.ClosedCB
+
+	c.conn.Opts.ClosedCB = func(c *nats.Conn) {
+		if closedCB != nil {
+			closedCB(c)
+		}
+
+		cancel(nil)
+	}
```

Comment:
Might be better to `defer cancel(nil)` here in case `closedCB` panics.
```suggestion
	c.conn.Opts.ClosedCB = func(c *nats.Conn) {
		defer cancel(nil)

		if closedCB != nil {
			closedCB(c)
		}
	}
```

Reply:
good point, thanks!